[AutoFill Debugging] Complete support for JavaScript-based text extraction filtering rules

whsieh opened the pull request:
#### 2da76b2d50b2a5bbf63de3b1b92378e0e242794f
<pre>
[AutoFill Debugging] Complete support for JavaScript-based text extraction filtering rules
<a href="https://bugs.webkit.org/show_bug.cgi?id=304280">https://bugs.webkit.org/show_bug.cgi?id=304280</a>
<a href="https://rdar.apple.com/166646585">rdar://166646585</a>

Reviewed by Aditya Keerthi.

Build on the infrastructure added in 303657@main, by requesting text extraction filtering rule data
and plumbing it to the web page before extracting text. See below for more details.

Tests: TextExtractionTests.FilteringRules

* Source/WebKit/Configurations/AllowedSPI.toml:
* Source/WebKit/Platform/spi/Cocoa/SafeBrowsingSPI.h:

Adjust this forward declaration to be consistent with the latest WIP API.

* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(createEmptyTextExtractionResult):
(-[WKWebView _extractDebugTextWithConfiguration:completionHandler:]):
(-[WKWebView _extractDebugTextWithConfigurationWithoutUpdatingFilterRules:completionHandler:]):

Move logic to extract debug text out into a separate method, and make the existing implementation of
`-_extractDebugTextWithConfiguration:completionHandler:` first ensure that the filtering rules have
been requested and updated (if available), before continuing the extraction.

* Source/WebKit/UIProcess/Cocoa/SafeBrowsingUtilities.h:
* Source/WebKit/UIProcess/Cocoa/SafeBrowsingUtilities.mm:
(WebKit::SafeBrowsingUtilities::listsForNamespace):

Adjust this to reflect the current state of the WIP API.

(WebKit::SafeBrowsingUtilities::namespacedCollectionForTextExtraction):
* Source/WebKit/UIProcess/Cocoa/TextExtraction/WKTextExtractionUtilities.h:
* Source/WebKit/UIProcess/Cocoa/TextExtraction/WKTextExtractionUtilities.mm:
(WebKit::requestTextExtractionFilterRuleData):

Add a helper function that requests raw list data using `listsForNamespace` above, and converts it
into a vector of `FilterRuleData`.

* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::hasTextExtractionFilterRules):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::hasTextExtractionFilterRules):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:
* Tools/TestWebKitAPI/Configurations/Base.xcconfig:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/TextExtractionTests.mm:
(TestWebKitAPI::overrideGetListsForNamespace):
(TestWebKitAPI::TEST(TextExtractionTests, FilteringRules)):

Add an API test to exercise this new functionality, by swizzling out the selector to contain
hard-coded rules.

Canonical link: <a href="https://commits.webkit.org/304580@main">https://commits.webkit.org/304580@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9c394f10da74a55e8625f414b4f554cb16aabef5

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/136020 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/8372 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/47294 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/143719 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/88992 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/569a4e73-1a80-41cc-a353-7c7f6606194b) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/137889 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/9041 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/8219 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/103943 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/88992 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/bdf3fdd3-c816-4170-ac19-6e587b88169c) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/138966 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/6555 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/121892 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/84821 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/88aeb91b-be32-47ca-a0e8-a1ac15ba42fd) 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/154/builds/6247 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac-debug~~](https://ews-build.webkit.org/#/builders/165/builds/3892 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe-cairo-libwebrtc](https://ews-build.webkit.org/#/builders/166/builds/4321 "Built successfully") | | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/115516 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/40086 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/146473 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/8056 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/40653 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/112298 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/8073 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/6760 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/112692 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/28594 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/6152 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/118195 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/62034 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/8105 "Built successfully") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/36263 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/7823 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/71662 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/8044 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/7901 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on Source/WebKit/UIProcess/Cocoa/TextExtraction/WKTextExtractionUtilities.mm:
```diff
@@ -285,6 +286,63 @@ static WKTextExtractionEventListenerTypes eventListenerTypes(OptionSet<TextExtra
+            if (keyIdentifierComponents.size() != 2)
+                continue;
+
+            auto ruleName = keyIdentifierComponents[0];
```
Maybe `.first()` instead of `[0]`? Would apply to multiple places in this patch.

whsieh replied:
Makes sense! Patched these three places:

```
diff --git a/Source/WebKit/UIProcess/Cocoa/TextExtraction/WKTextExtractionUtilities.mm b/Source/WebKit/UIProcess/Cocoa/TextExtraction/WKTextExtractionUtilities.mm
index 7a773969fa68..c1ffd9cbc451 100644
--- a/Source/WebKit/UIProcess/Cocoa/TextExtraction/WKTextExtractionUtilities.mm
+++ b/Source/WebKit/UIProcess/Cocoa/TextExtraction/WKTextExtractionUtilities.mm
@@ -303,7 +303,7 @@ void requestTextExtractionFilterRuleData(CompletionHandler<void(Vector<TextExtra
             if (keyIdentifierComponents.size() != 2)
                 continue;

-            auto ruleName = keyIdentifierComponents[0];
+            auto ruleName = keyIdentifierComponents.first();
             if (ruleName.isEmpty())
                 continue;

@@ -322,10 +322,10 @@ void requestTextExtractionFilterRuleData(CompletionHandler<void(Vector<TextExtra
                 ensureRuleData().urlPatternString = [&] -> String {
                     auto domainRules = makeVector<String>(data[nsKeyIdentifier]);
                     if (domainRules.size() == 1) {
-                        if (domainRules[0] == ".*"_s)
+                        if (domainRules.first() == ".*"_s)
                             return { };

-                        return domainRules[0];
+                        return domainRules.first();
                     }

                     return makeStringByJoining(WTF::map(WTFMove(domainRules), [](auto&& group) {
```

whsieh committed:
[AutoFill Debugging] Complete support for JavaScript-based text extraction filtering rules
https://bugs.webkit.org/show_bug.cgi?id=304280
rdar://166646585

Reviewed by Aditya Keerthi.

Build on the infrastructure added in 303657@main, by requesting text extraction filtering rule data
and plumbing it to the web page before extracting text. See below for more details.

Tests: TextExtractionTests.FilteringRules

* Source/WebKit/Configurations/AllowedSPI.toml:
* Source/WebKit/Platform/spi/Cocoa/SafeBrowsingSPI.h:

Adjust this forward declaration to be consistent with the latest WIP API.

* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(createEmptyTextExtractionResult):
(-[WKWebView _extractDebugTextWithConfiguration:completionHandler:]):
(-[WKWebView _extractDebugTextWithConfigurationWithoutUpdatingFilterRules:completionHandler:]):

Move logic to extract debug text out into a separate method, and make the existing implementation of
`-_extractDebugTextWithConfiguration:completionHandler:` first ensure that the filtering rules have
been requested and updated (if available), before continuing the extraction.

* Source/WebKit/UIProcess/Cocoa/SafeBrowsingUtilities.h:
* Source/WebKit/UIProcess/Cocoa/SafeBrowsingUtilities.mm:
(WebKit::SafeBrowsingUtilities::listsForNamespace):

Adjust this to reflect the current state of the WIP API.

(WebKit::SafeBrowsingUtilities::namespacedCollectionForTextExtraction):
* Source/WebKit/UIProcess/Cocoa/TextExtraction/WKTextExtractionUtilities.h:
* Source/WebKit/UIProcess/Cocoa/TextExtraction/WKTextExtractionUtilities.mm:
(WebKit::requestTextExtractionFilterRuleData):

Add a helper function that requests raw list data using `listsForNamespace` above, and converts it
into a vector of `FilterRuleData`.

* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::hasTextExtractionFilterRules):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::hasTextExtractionFilterRules):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:
* Tools/TestWebKitAPI/Configurations/Base.xcconfig:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/TextExtractionTests.mm:
(TestWebKitAPI::overrideGetListsForNamespace):
(TestWebKitAPI::TEST(TextExtractionTests, FilteringRules)):

Add an API test to exercise this new functionality, by swizzling out the selector to contain
hard-coded rules.

Canonical link: https://commits.webkit.org/304580@main
